sympy/polys/rootoftools.py

```diff
@@ -271,135 +321,69 @@ def _reals_sorted(cls, reals):
+                    complexes.append((u, f, k))
+            else:
+                # refine until all but nimag xbounds are neg or pos
+                potential_imag = list(range(len(sifted[f])))
+                while True:
+                    assert len(potential_imag) > 1
+                    done = []
+                    for i in potential_imag:
+                        u, f, k = sifted[f][i]
+                        if u.ax*u.bx > 0:
+                            done.append(i)
+                        else:
+                            u = u._inner_refine()
+                            sifted[f][i] = u, f, k
+                    for i in done:
+                        potential_imag.remove(i)
+                    if len(potential_imag) == nimag:
+                        break
+                complexes.extend(sifted[f])
+        return complexes
 
     @classmethod
     def _refine_complexes(cls, complexes):
         """return complexes such that no bounding rectangles of non-conjugate
         roots would intersect if slid horizontally or vertically/
         """
-        while complexes:  # break when all are distinct
-            # get the intervals pairwise-disjoint.
-            # If rectangles were drawn around the coordinates of the bounding
-            # rectangles, no rectangles would intersect after this procedure.
-            for i, (u, f, k) in enumerate(complexes):
-                for j, (v, g, m) in enumerate(complexes[i + 1:]):
-                    u, v = u.refine_disjoint(v)
-                    complexes[i + j + 1] = (v, g, m)
-
-                complexes[i] = (u, f, k)
-            # Although there are no intersecting rectangles, a given rectangle
-            # might intersect another when slid horizontally. We have to refine
-            # intervals until this is not true so we can sort the roots
-            # unambiguously. Since complex roots come in conjugate pairs, we
-            # will always have 2 rectangles above each other but we should not
-            # have more than that.
-            N = len(complexes)//2 - 1
-            # check x (real) parts: there must be N + 1 disjoint x ranges, i.e.
-            # the first one must be different from N others
-            uu = set([(u.ax, u.bx) for u, _, _ in complexes])
-            u = uu.pop()
-            if sum([u[1] <= v[0] or v[1] <= u[0] for v in uu]) < N:
-                # refine
-                for i, (u, f, k) in enumerate(complexes):
-                    u = u._inner_refine()
-                    complexes[i] = u, f, k
-            else:
-                # intervals with identical x-values have disjoint y-values or
-                # else they would not be disjoint so there is no need for
-                # further checks
-                break
+        # get the intervals pairwise-disjoint.
+        # If rectangles were drawn around the coordinates of the bounding
+        # rectangles, no rectangles would intersect after this procedure.
```

Comment:
Oh, I bet they will...

Reply:
This is the task of `refine_disjoint`: refine the boundaries until either the x range or y range of self is distinct from other. When this condition is met the rectangles can no longer intersect. This does not mean that they will not overlap *when slid*, however. That's why being disjoint is not sufficient for sorting. (I am only guessing that that is what you are thinking is being claimed by this comment. It would help if you actually give an example so we have something concrete to talk about.)

Reply:
Oh, indead, I wrongly interpret this as being pairwise-disjoint in real part.  Sorry.